src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_Extra.cs

```diff
@@ -1707,6 +1707,243 @@ static void Main()
             """, graph, symbol);
     }
 
+    private string CreateCustomListDefinition(string constructors)
+    {
+        return $$"""
+            namespace System.Collections.Generic
+            {
+                public class List<T> : IList<T>
+                {
+                    {{constructors}}
+            
+                    public T this[int index] { get => throw null; set => throw null; }
+                    public int Count => throw null;
+                    public bool IsReadOnly => throw null;
+                    public void Add(T item) => throw null;
+                    public void Clear() => throw null;
+                    public bool Contains(T item) => throw null;
+                    public void CopyTo(T[] array, int arrayIndex) => throw null;
+                    public IEnumerator<T> GetEnumerator() => throw null;
+                    public int IndexOf(T item) => throw null;
+                    public void Insert(int index, T item) => throw null;
+                    public bool Remove(T item) => throw null;
+                    public void RemoveAt(int index) => throw null;
+                    IEnumerator IEnumerable.GetEnumerator() => throw null;
+                    public T[] ToArray() => throw null;
+                }
+            }
+            """;
+    }
+
+    [Theory]
+    [InlineData("IList<System.Int32>")]
+    [InlineData("ICollection<System.Int32>")]
+    public void IList_With_DifferentCapacityName(string typeName)
+    {
+        var source = $$"""
+            using System.Collections.Generic;
+
+            class Program
+            {
+                static void Main()
+                {
+                    {{typeName}} x = [with(), 1, 2, 3];
+                    {{typeName}} y = [with(cap: 6), 1, 2, 3];
+                }
+            }
+            """;
+
+        var compilation = CreateCompilation([
+            source,
+            CreateCustomListDefinition("""public List() { } public List(int cap) { }""")],
+            assemblyName: "MyAssembly");
+        CompileAndVerify(compilation).VerifyDiagnostics();
```

Comment:
>CompileAndVerify(compilation).VerifyDiagnostics(); [](http://example.com/codeflow?start=8&length=50)

I think we would want to execute the code and observe what constructor is getting called. This can be achieved by printing something distinct from each of the constructors. Obviously, we wouldn't want the `Add` method to throw as well, simply doing nothing in the method would be sufficient for the purpose of these tests.

Reply:
done:  35f2c09b3d2